integration_tests/lock_test.go

```diff
@@ -1248,3 +1254,269 @@ func (s *testLockWithTiKVSuite) TestPrewriteCheckForUpdateTS() {
+	http.DefaultClient.CloseIdleConnections()
+
+	if len(recoverConfigs) > 0 {
+		// Sleep for a while to ensure the new configs are applied.
+		time.Sleep(time.Second)
+	}
+
+	return func() {
+		for _, item := range recoverConfigs {
+			err = setCfg(item.url, item.name, item.value)
+			if err != nil {
+				t.Logf("failed to recover config for store at %s: %v", item.url, err)
+			}
+		}
+
+		// Prevent goleak from complaining about its internal connections.
+		http.DefaultClient.CloseIdleConnections()
+	}
+}
+
+func (s *testLockWithTiKVSuite) TestStaleForcePessimisticLockNotCommitted() {
+	// See https://github.com/tikv/tikv/issues/14551 for detail about the issue this test is supposed to cover.
+	if !*withTiKV {
+		// The issue only exists in TiKV's storage model. Unnecessary to run with mocks.
+		return
+	}
+
+	recoverFunc := s.setTiKVConfig("pessimistic-txn.in-memory", false)
+	defer recoverFunc()
+
+	ctx := context.WithValue(context.Background(), util.SessionID, uint64(1))
+
+	var success bool
+	for iteration := 0; iteration < 20; iteration++ {
+		success = func() bool {
+			// Test several times repeatedly for greater possibility to reproduce the issue if it exists.
+			s.T().Logf("TestStaleForcePessimisticLockNotCommitted starts iteration %d", iteration)
+
+			// Add prefix to the key to make that more likely to locate in the same region.
+			pk := []byte(fmt.Sprintf("TestStaleForcePessimisticLockNotCommitted-k%d-p", iteration))
+			key1 := []byte(fmt.Sprintf("TestStaleForcePessimisticLockNotCommitted-k%d-1", iteration))
+			key2 := []byte(fmt.Sprintf("TestStaleForcePessimisticLockNotCommitted-k%d-2", iteration))
+
+			v1 := []byte("v1")
+			v2 := []byte("v2")
+
+			// Remove the keys if they already exists
+			{
+				t, err := s.store.Begin()
+				s.NoError(err)
+				s.NoError(t.Delete(pk))
+				s.NoError(t.Delete(key1))
+				s.NoError(t.Delete(key2))
+				s.NoError(t.Commit(ctx))
+			}
+
+			t1, err := s.store.Begin()
+			s.NoError(err)
+			t1.SetPessimistic(true)
+			t1.SetEnableAsyncCommit(true)
+			// Make t1's primary determined
+			lockCtx := kv.NewLockCtx(t1.StartTS(), 200, time.Now())
+			s.NoError(t1.LockKeys(ctx, lockCtx, pk))
+			s.NoError(t1.Set(pk, v1))
+
+			simulatedCommiter, err := t1.NewCommitter(1)
+			s.NoError(err)
+
+			t2, err := s.store.Begin()
+			s.NoError(err)
+			t2.SetEnableAsyncCommit(true)
+			s.NoError(t2.Set(key1, v2))
+			s.NoError(t2.Set(key2, v2))
+
+			t1.StartAggressiveLocking()
+			lockTime := time.Now()
+			lockCtx = kv.NewLockCtx(t1.StartTS(), 200, lockTime)
+			s.NoError(t1.LockKeys(ctx, lockCtx, key1))
+			t1.CancelAggressiveLocking(ctx)
+			// Wait the pessimistic lock of t1 on key1 to be rolled back.
+			s.checkIsKeyLocked(key1, false)
+
+			// Try to make transaction t1 and t2 commit at the same commitTS.
+			// First, block at finishing getting minCommitTS from PD and before prewriting
+			s.NoError(failpoint.Enable("tikvclient/beforePrewrite", "pause"))
+			resCh := make(chan error)
```

Comment:
Is it better to use `resCh := make(chan error, 2)` so they two results would not block each other on the result channel?